File: services/manager/grpc/word.go
package grpc

import (
	"context"

	"github.com/Amakuchisan/tsuginiyomu/services/manager/manager"
	pb "github.com/Amakuchisan/tsuginiyomu/services/manager/pb/manager"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
)

// CreateWord は新規単語の登録を行う
func (s *Server) CreateWord(ctx context.Context, in *pb.CreateWordRequest) (*pb.CreateWordReply, error) {
	for _, article := range in.Article {
		art, err := s.manager.GetArticle(ctx, article.Url)
		if err != nil {
			return &pb.CreateWordReply{Created: false}, err
		}
		_, err = s.manager.CreateWord(ctx, article.WordCount)
		if err != nil {
			return &pb.CreateWordReply{Created: false}, err
		}
		err = s.manager.CreateArticleWord(ctx, art.ID, article.WordCount)
		if err != manager.ErrAlreadyRegistered {
			if err != nil {
				return &pb.CreateWordReply{Created: false}, nil
			}
		}
	}

	return &pb.CreateWordReply{Created: true}, nil
}

// GetWord は新規単語の検索を行い, 結果を返す
func (s *Server) GetWord(ctx context.Context, in *pb.GetWordRequest) (*pb.GetWordReply, error) {
	wordCount, err := s.manager.GetWord(ctx, in.HatenaID)
	if err != nil {
		if err == manager.ErrInvalidArgument {
			return nil, status.Error(codes.InvalidArgument, "invalid argument")
		}
		return nil, err
	}
	return &pb.GetWordReply{WordCount: wordCount}, nil
}

File: services/web/manager/word.py
import grpc

import pb.manager.manager_pb2 as manager_pb2
import pb.manager.manager_pb2_grpc as manager_pb2_grpc

# 辞書型で返却
def find_word(hatena_id: str) -> dict[str, int]:
    with grpc.insecure_channel('manager:50051') as channel:
        stub = manager_pb2_grpc.ManagerStub(channel)
        response = stub.GetWord(manager_pb2.GetWordRequest(hatenaID=hatena_id))
    return response.wordCount

def create(article: manager_pb2.Article()):
    with grpc.insecure_channel('manager:50051') as channel:
        stub = manager_pb2_grpc.ManagerStub(channel)
        response = stub.CreateWord(createRequest(article))

def createRequest(article: manager_pb2.Article()) -> manager_pb2.CreateWordRequest():
    req = manager_pb2.CreateWordRequest()
    req.article.extend(article)
    return req
